File: apps/ios/Voiid/Voiid/Games/GameSetupSheet.swift
//
//  GameSetupSheet.swift
//  Voiid
//
//  The one step between tapping a game and playing it.
//

import SwiftUI

struct GameSetupSheet: View {

    let game: Game
    var onStart: (GameMode) -> Void = { _ in }

    @Environment(\.dismiss) private var dismiss

    @State private var selected: GameMode?
    /// The thumb's own position, which is continuous — `selected` is the discrete mode it
    /// resolves to. Keeping them apart is what lets the thumb sit between two stops.
    @State private var sliderValue: Double = 0
    @State private var selectedDetent: PresentationDetent = .medium

    private var modes: [GameMode] { GameMode.modes(for: game.id) }

    var body: some View {
        ZStack {
            VoiidColor.background.ignoresSafeArea()

            VStack(spacing: 0) {
                banner

                ScrollView(.vertical, showsIndicators: false) {
                    VStack(alignment: .leading, spacing: VoiidSpacing.lg) {
                        VStack(alignment: .leading, spacing: VoiidSpacing.sm) {
                            Text("How do you want to play?")
                                .font(VoiidFont.rounded(17, .bold))
                                .foregroundColor(VoiidColor.textPrimary)

                            difficultySlider
                        }
                        .frame(maxWidth: .infinity, alignment: .leading)

                        rulesSection
                            .frame(maxWidth: .infinity, alignment: .leading)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, VoiidSpacing.md)
                    .padding(.top, VoiidSpacing.md)
                    .padding(.bottom, VoiidSpacing.md)
                }
                .scrollBounceBehavior(.basedOnSize, axes: .horizontal)

                footer
            }
            .frame(maxWidth: .infinity)
        }
        .presentationDetents([.medium, .large], selection: $selectedDetent)
        .presentationDragIndicator(.visible)
        .presentationCornerRadius(28)
        .onAppear {
            if selected == nil { selected = modes.first }
            sliderValue = Double(modes.firstIndex { $0.id == selected?.id } ?? 0)
        }
    }

    private var banner: some View {
        ZStack(alignment: .bottomLeading) {
            LinearGradient(colors: [game.tintA, game.tintB.opacity(0.75)],
                           startPoint: .topLeading, endPoint: .bottomTrailing)

            Image(systemName: game.symbol)
                .font(.system(size: 92, weight: .medium))
                .foregroundColor(.white.opacity(0.16))
                .offset(x: 170, y: 20)

            VStack(alignment: .leading, spacing: 3) {
                Text(game.title)
                    .font(VoiidFont.rounded(22, .bold))
                    .foregroundColor(.white)
                    .shadow(color: .black.opacity(0.35), radius: 5, y: 2)

                if let tagline = GameRules.tagline(for: game.id) {
                    Text(tagline)
                        .font(VoiidFont.rounded(12, .medium))
                        .foregroundColor(.white.opacity(0.88))
                        .lineLimit(1)
                }

                HStack(spacing: 10) {
                    if let players = game.players { chip("person.2.fill", players) }
                    if let minutes = game.minutes { chip("clock", minutes) }
                }
            }
            .padding(VoiidSpacing.md)
        }
        .frame(maxWidth: .infinity)
        .frame(height: 112)
        .clipped()
        // Runs to the sheet's own edges and rounds to match its 28pt corners, so the
        // artwork meets the sheet instead of sitting in it as a square inset panel.
        .clipShape(
            UnevenRoundedRectangle(
                topLeadingRadius: 28, bottomLeadingRadius: 0,
                bottomTrailingRadius: 0, topTrailingRadius: 28,
                style: .continuous
            )
        )
    }

    @ViewBuilder
    private var rulesSection: some View {
        let rules = GameRules.lines(for: game.id)
        if !rules.isEmpty {
            VStack(alignment: .leading, spacing: VoiidSpacing.sm) {
                Text("Rules & Objectives")
                    .font(VoiidFont.rounded(16.5, .bold))
                    .foregroundColor(VoiidColor.textPrimary)

                VStack(spacing: 12) {
                    ForEach(rules) { rule in
                        HStack(alignment: .top, spacing: 12) {
                            Image(systemName: rule.icon)
                                .font(.system(size: 14, weight: .semibold))
                                .foregroundColor(VoiidColor.accent)
                                .frame(width: 26, height: 26)
                                .background(Circle().fill(VoiidColor.accentSoft))

                            Text(rule.text)
                                .font(VoiidFont.rounded(13))
                                .foregroundColor(VoiidColor.textPrimary.opacity(0.92))
                                .lineSpacing(2)
                                .frame(maxWidth: .infinity, alignment: .leading)
                        }
                        .frame(maxWidth: .infinity, alignment: .leading)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(VoiidSpacing.md)
                .background(VoiidColor.surfaceCard)
                .clipShape(RoundedRectangle(cornerRadius: VoiidRadius.md, style: .continuous))
                .overlay(
                    RoundedRectangle(cornerRadius: VoiidRadius.md, style: .continuous)
                        .stroke(VoiidColor.divider, lineWidth: 1)
                )
            }
        }
    }

    private func chip(_ icon: String, _ text: String) -> some View {
        HStack(spacing: 4) {
            Image(systemName: icon)
                .font(.system(size: 9.5))
            Text(text)
                .font(VoiidFont.rounded(11.5, .medium))
        }
        .foregroundColor(.white.opacity(0.92))
    }

    /// Difficulty as one track rather than a stack of radio rows.
    ///
    /// The modes are ordered and mutually exclusive — easy, moderate, hard — which is a
    /// magnitude, not a set of unrelated options. A slider says that in one control and one
    /// line of text, where three cards said it in three boxes competing for the same glance.
    private var difficultySlider: some View {
        let index = modes.firstIndex { $0.id == selected?.id } ?? 0
        let mode = modes[min(index, modes.count - 1)]

        return VStack(alignment: .leading, spacing: VoiidSpacing.md) {
            HStack(spacing: 10) {
                Image(systemName: mode.icon)
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundColor(VoiidColor.accent)
                    .frame(width: 22)
                    .contentTransition(.symbolEffect(.replace))

                VStack(alignment: .leading, spacing: 1) {
                    Text(mode.title)
                        .font(VoiidFont.rounded(15, .semibold))
                        .foregroundColor(VoiidColor.textPrimary)
                    Text(mode.detail)
                        .font(VoiidFont.rounded(12.5))
                        .foregroundColor(VoiidColor.textSecondary)
                }
                Spacer(minLength: 0)
            }
            // Reserved so the card does not resize as the two lines change length.
            .frame(height: 40, alignment: .leading)
            .animation(.easeOut(duration: 0.18), value: mode.id)

            // CONTINUOUS WHILE DRAGGING, SETTLING ON RELEASE.
            //
            // `step: 1` made the thumb teleport between the three stops: it cannot rest
            // between them, so the control stopped tracking the thumb and the motion read as
            // broken rather than as snapping. Here the thumb follows the finger exactly, the
            // selection updates as it crosses each stop, and only on release does it glide to
            // the chosen one — which is what makes the travel feel smooth and still land on a
            // real value.
            Slider(
                value: $sliderValue,
                in: 0...Double(max(modes.count - 1, 1)),
                onEditingChanged: { editing in
                    guard !editing else { return }
                    withAnimation(.spring(response: 0.28, dampingFraction: 0.78)) {
                        sliderValue = Double(index)
                    }
                }
            )
            .tint(VoiidColor.accent)
            .accessibilityLabel("Difficulty")
            .accessibilityValue(mode.title)
            // Crossing a stop mid-drag picks it, so the description and icon above update
            // under the thumb rather than only once the finger lifts.
            .onChange(of: sliderValue) { _, new in
                let step = min(max(Int(new.rounded()), 0), modes.count - 1)
                guard step != index else { return }
                Haptics.selection()
                selected = modes[step]
            }

            HStack {
                ForEach(Array(modes.enumerated()), id: \.element.id) { offset, m in
                    Text(m.shortLabel)
                        .font(VoiidFont.rounded(11.5, offset == index ? .bold : .medium))
                        .foregroundColor(offset == index ? VoiidColor.accent
                                                         : VoiidColor.textSecondary)
                    if offset < modes.count - 1 { Spacer(minLength: 0) }
                }
            }
        }
        .padding(VoiidSpacing.md)
        .background(VoiidColor.surfaceCard)
        .clipShape(RoundedRectangle(cornerRadius: VoiidRadius.md, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: VoiidRadius.md, style: .continuous)
            .stroke(VoiidColor.divider, lineWidth: 1))
    }

    private var footer: some View {
        VStack(spacing: VoiidSpacing.sm) {
            Button {
                guard let selected else { return }
                Haptics.success()
                onStart(selected)
                dismiss()
            } label: {
                HStack(spacing: 8) {
                    Image(systemName: "play.fill")
                        .font(.system(size: 14, weight: .bold))
                    Text("Start game")
                        .font(VoiidFont.rounded(16.5, .semibold))
                }
                .foregroundColor(VoiidColor.textOnAccent)
                .frame(maxWidth: .infinity)
                .frame(height: 52)
                .background(RoundedRectangle(cornerRadius: VoiidRadius.lg,
                                             style: .continuous)
                    .fill(VoiidColor.accent))
            }
            .buttonStyle(PressableButtonStyle())
            .disabled(selected == nil)
            .opacity(selected == nil ? 0.45 : 1)

            Button("Not now") { dismiss() }
                .font(VoiidFont.rounded(14.5, .semibold))
                .foregroundColor(VoiidColor.textSecondary)
        }
        .padding(.horizontal, VoiidSpacing.md)
        .padding(.top, VoiidSpacing.sm)
        .padding(.bottom, VoiidSpacing.lg)
        .background(.bar)
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/games/GameSetupSheet.kt
package com.voiid.app.main.games

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.outlined.Bolt
import androidx.compose.material.icons.outlined.Memory
import androidx.compose.material.icons.outlined.Palette
import androidx.compose.material.icons.outlined.Spa
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.roundToInt
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont
import com.voiid.app.ui.theme.VoiidRadius
import com.voiid.app.ui.theme.VoiidSpacing

/**
 * Game setup sheet: Offline bot & practice mode with clear difficulty selection.
 * Full parity twin of iOS GameSetupSheet.swift.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun GameSetupSheet(
    gameName: String,
    slug: String = "",
    onPlayFriend: () -> Unit = {},
    onPlayBot: ((BotDifficulty, Float) -> Unit)? = null,
    onCustomise: (() -> Unit)? = null,
    onDismiss: () -> Unit,
) {
    var selectedDifficulty by remember { mutableStateOf(BotDifficulty.EASY) }
    // The thumb's own position, which is continuous — `selectedDifficulty` is the discrete
    // mode it resolves to. Keeping them apart lets the thumb sit between two stops.
    var sliderValue by remember { mutableFloatStateOf(0f) }
    var dragging by remember { mutableStateOf(false) }

    com.voiid.app.ui.components.VoiidSheet(
        visible = true,
        onDismiss = onDismiss,
        detents = listOf(
            com.voiid.app.ui.components.VoiidDetent.Medium,
            com.voiid.app.ui.components.VoiidDetent.Large,
        ),
        initialDetentIndex = 0,
        showHandle = true,
    ) {
        Column(
            Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = VoiidSpacing.md)
                .padding(bottom = VoiidSpacing.xl),
            verticalArrangement = Arrangement.spacedBy(VoiidSpacing.md),
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(
                    gameName,
                    color = VoiidColor.textPrimary,
                    style = VoiidFont.rounded(22, FontWeight.Bold),
                )
                GameRules.tagline(slug)?.let { tagline ->
                    Text(
                        tagline,
                        color = VoiidColor.textSecondary,
                        style = VoiidFont.rounded(13, FontWeight.Normal),
                    )
                }
            }

            // Rules card
            val rules = GameRules.lines(slug)
            if (rules.isNotEmpty()) {
                Column(
                    Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(VoiidRadius.md))
                        .background(VoiidColor.surfaceCard)
                        .padding(VoiidSpacing.sm + 4.dp),
                    verticalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    rules.forEach { line ->
                        Row(
                            verticalAlignment = Alignment.Top,
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                        ) {
                            Icon(
                                line.icon,
                                contentDescription = null,
                                tint = VoiidColor.primary,
                                modifier = Modifier.size(15.dp),
                            )
                            Text(
                                line.text,
                                color = VoiidColor.textSecondary,
                                style = VoiidFont.rounded(13, FontWeight.Normal),
                            )
                        }
                    }
                }
            }

            if (onCustomise != null) {
                Row(
                    Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(VoiidRadius.md))
                        .background(VoiidColor.surfaceCard)
                        .clickable { onCustomise() }
                        .padding(horizontal = VoiidSpacing.md, vertical = 12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(VoiidSpacing.md),
                ) {
                    Icon(
                        Icons.Outlined.Palette,
                        contentDescription = null,
                        tint = VoiidColor.primary,
                        modifier = Modifier.size(20.dp),
                    )
                    Column(Modifier.weight(1f)) {
                        Text(
                            "Your snake",
                            color = VoiidColor.textPrimary,
                            style = VoiidFont.rounded(14, FontWeight.SemiBold),
                        )
                        Text(
                            "Pick a skin or a colour",
                            color = VoiidColor.textSecondary,
                            style = VoiidFont.rounded(12, FontWeight.Normal),
                        )
                    }
                }
            }

            Text(
                "How do you want to play?",
                color = VoiidColor.textPrimary,
                style = VoiidFont.rounded(16, FontWeight.Bold),
            )

            // Difficulty as one track rather than three radio rows: the modes are ordered
            // and mutually exclusive, which is a magnitude, not a set of unrelated options.
            // Mirrors iOS GameSetupSheet.swift.
            val steps = listOf(
                Triple(BotDifficulty.EASY, Icons.Outlined.Spa, "Easy"),
                Triple(BotDifficulty.MODERATE, Icons.Outlined.Memory, "Moderate"),
                Triple(BotDifficulty.HARD, Icons.Outlined.Bolt, "Hard"),
            )
            val index = steps.indexOfFirst { it.first == selectedDifficulty }.coerceAtLeast(0)
            val (_, stepIcon, _) = steps[index]
            val title = when (selectedDifficulty) {
                BotDifficulty.EASY -> if (slug == "snake") "Easy Mode (Practice)" else "Easy Mode (vs Bots)"
                BotDifficulty.MODERATE -> if (slug == "snake") "Standard Arena" else "Moderate Mode (vs Bots)"
                BotDifficulty.HARD -> if (slug == "snake") "Hardcore Arena" else "Hard Mode (vs Bots)"
            }
            val detail = when (selectedDifficulty) {
                BotDifficulty.EASY -> if (slug == "carrom") "You play white, the bot plays black" else if (slug == "snake") "Calm arena with fewer bots, relaxed growth" else "Relaxed bots, gentle moves, easy practice"
                BotDifficulty.MODERATE -> if (slug == "carrom") "You play white, the bot plays black" else if (slug == "snake") "Competitive 8 bots, balanced speed" else "Standard balanced match against 3 bots"
                BotDifficulty.HARD -> if (slug == "carrom") "You play white, the bot plays black" else if (slug == "snake") "Fast paced, aggressive hunting bots" else "Aggressive bots that cut and race"
            }

            Column(
                Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(VoiidRadius.md))
                    .background(VoiidColor.surfaceCard)
                    .padding(VoiidSpacing.md),
                verticalArrangement = Arrangement.spacedBy(VoiidSpacing.sm),
            ) {
                Row(
                    // Reserved so the card does not resize as the two lines change length.
                    Modifier.fillMaxWidth().height(44.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(10.dp),
                ) {
                    Icon(
                        stepIcon,
                        contentDescription = null,
                        tint = VoiidColor.primary,
                        modifier = Modifier.size(20.dp),
                    )
                    Column(Modifier.weight(1f)) {
                        Text(
                            title,
                            color = VoiidColor.textPrimary,
                            style = VoiidFont.rounded(14, FontWeight.SemiBold),
                        )
                        Text(
                            detail,
                            color = VoiidColor.textSecondary,
                            style = VoiidFont.rounded(12, FontWeight.Normal),
                        )
                    }
                }

                // CONTINUOUS WHILE DRAGGING, SETTLING ON RELEASE.
                //
                // `steps` made the thumb teleport between the three stops: it cannot rest
                // between them, so the control stopped tracking the finger and the motion
                // read as broken rather than as snapping. Here the thumb follows exactly,
                // the selection updates as it crosses each stop, and on release it animates
                // to the chosen one. Mirrors iOS GameSetupSheet.swift.
                val settled by androidx.compose.animation.core.animateFloatAsState(
                    targetValue = index.toFloat(),
                    animationSpec = androidx.compose.animation.core.spring(
                        dampingRatio = 0.78f,
                        stiffness = androidx.compose.animation.core.Spring.StiffnessMediumLow,
                    ),
                    label = "difficultySettle",
                )
                androidx.compose.material3.Slider(
                    value = if (dragging) sliderValue else settled,
                    onValueChange = { raw ->
                        dragging = true
                        sliderValue = raw
                        val step = raw.roundToInt().coerceIn(0, steps.lastIndex)
                        if (steps[step].first != selectedDifficulty) {
                            selectedDifficulty = steps[step].first
                        }
                    },
                    onValueChangeFinished = { dragging = false },
                    valueRange = 0f..steps.lastIndex.toFloat(),
                    colors = androidx.compose.material3.SliderDefaults.colors(
                        thumbColor = VoiidColor.primary,
                        activeTrackColor = VoiidColor.primary,
                    ),
                )

                Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                    steps.forEachIndexed { offset, (_, _, label) ->
                        Text(
                            label,
                            color = if (offset == index) VoiidColor.primary else VoiidColor.textSecondary,
                            style = VoiidFont.rounded(
                                11.5f,
                                if (offset == index) FontWeight.Bold else FontWeight.Medium,
                            ),
                        )
                    }
                }
            }

            Spacer(Modifier.height(4.dp))

            // Start Game Button
            Box(
                Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(VoiidRadius.lg))
                    .background(VoiidColor.primary)
                    .clickable {
                        onPlayBot?.invoke(selectedDifficulty, selectedDifficulty.skill)
                    }
                    .padding(vertical = 15.dp),
                contentAlignment = Alignment.Center,
            ) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    Icon(
                        Icons.Default.PlayArrow,
                        contentDescription = null,
                        tint = VoiidColor.textOnPrimary,
                        modifier = Modifier.size(18.dp),
                    )
                    Text(
                        "Start game",
                        color = VoiidColor.textOnPrimary,
                        style = VoiidFont.rounded(16, FontWeight.SemiBold),
                    )
                }
            }
        }
    }
}

@Composable
private fun DifficultyOptionRow(
    title: String,
    detail: String,
    icon: ImageVector,
    isSelected: Boolean,
    onClick: () -> Unit,
) {
    Row(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(VoiidRadius.md))
            .background(if (isSelected) VoiidColor.primary.copy(alpha = 0.14f) else VoiidColor.surfaceCard)
            .border(
                width = if (isSelected) 1.5.dp else 1.dp,
                color = if (isSelected) VoiidColor.primary else VoiidColor.divider,
                shape = RoundedCornerShape(VoiidRadius.md),
            )
            .clickable { onClick() }
            .padding(horizontal = VoiidSpacing.md, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(VoiidSpacing.md),
    ) {
        Icon(
            icon,
            contentDescription = null,
            tint = if (isSelected) VoiidColor.primary else VoiidColor.textSecondary,
            modifier = Modifier.size(22.dp),
        )

        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(
                title,
                color = VoiidColor.textPrimary,
                style = VoiidFont.rounded(14, FontWeight.SemiBold),
            )
            Text(
                detail,
                color = VoiidColor.textSecondary,
                style = VoiidFont.rounded(12, FontWeight.Normal),
            )
        }

        // Selection indicator
        Box(
            Modifier
                .size(20.dp)
                .border(
                    width = if (isSelected) 2.dp else 1.5.dp,
                    color = if (isSelected) VoiidColor.primary else VoiidColor.textSecondary,
                    shape = CircleShape,
                ),
            contentAlignment = Alignment.Center,
        ) {
            if (isSelected) {
                Box(
                    Modifier
                        .size(10.dp)
                        .clip(CircleShape)
                        .background(VoiidColor.primary)
                )
            }
        }
    }
}
